{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# GarNet\n",
    "\n",
    "There are two steps to GarNet: \n",
    "- **mapping** a BED file of peaks to TF motif/gene complexes \n",
    "- **regressing** those motifs' binding affinities against gene expression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "from GarNet import map_peaks, TF_regression\n",
    "\n",
    "garnet_file = \"../example_data/hg19/hg19.garnet.pickle\"\n",
    "\n",
    "peaks = \"../example_data/example_peaks/wgEncodeUWDukeDnaseA549.fdr01peaks.hg19.bed\"\n",
    "# peaks = \"../example/example_peaks/A549_FOXA1_broadPeak.bed\"\n",
    "# peaks = \"../example/example_peaks/JUNB_ChIP_A549.bed\"\n",
    "# peaks = \"../example/example_peaks/CTCF_ChIP_A549.bed\"\n",
    "# peaks = [\"../example/example_peaks/A549_FOXA1_broadPeak.bed\", \"../example/example_peaks/JUNB_ChIP_A549.bed\", \"../example/example_peaks/CTCF_ChIP_A549.bed\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Map Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "02:10:32 - GarNet: INFO - Mapping peaks against genome from garnet-file ../example_data/hg19/hg19.garnet.pickle\n",
      "02:10:32 - GarNet: INFO - Unpacking garnet file (this can take a while)... \n",
      "02:48:14 - GarNet: INFO - Constructing representation of peaks file ../example_data/example_peaks/wgEncodeUWDukeDnaseA549.fdr01peaks.hg19.bed...\n",
      "02:48:14 - GarNet: INFO -   - Peaks file does not seem to have been generated by pickle, proceeding to parse...\n",
      "02:48:15 - GarNet: INFO -   - Parse complete, constructing IntervalTrees...\n",
      "02:48:23 - GarNet: INFO - Computing intersection of peaks with reference...\n",
      "02:48:23 - GarNet: INFO - Computing intersection operation of IntervalTrees for each chromosome...\n"
     ]
    }
   ],
   "source": [
    "df = map_peaks(peaks, garnet_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>motifStart</th>\n",
       "      <th>motifEnd</th>\n",
       "      <th>motifName</th>\n",
       "      <th>motifScore</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>peakName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>10009681</td>\n",
       "      <td>10009687</td>\n",
       "      <td>MZF1_1-4</td>\n",
       "      <td>0.395137</td>\n",
       "      <td>WWC3</td>\n",
       "      <td>9983794</td>\n",
       "      <td>10112518</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039141</td>\n",
       "      <td>25039160</td>\n",
       "      <td>REST</td>\n",
       "      <td>0.015574</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039207</td>\n",
       "      <td>25039218</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>0.029736</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039140</td>\n",
       "      <td>25039161</td>\n",
       "      <td>NRSE</td>\n",
       "      <td>0.019347</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039140</td>\n",
       "      <td>25039161</td>\n",
       "      <td>REST</td>\n",
       "      <td>0.029062</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  motifStart  motifEnd motifName  motifScore geneName  geneStart  \\\n",
       "0  chrX    10009681  10009687  MZF1_1-4    0.395137     WWC3    9983794   \n",
       "1  chrX    25039141  25039160      REST    0.015574      ARX   25021812   \n",
       "2  chrX    25039207  25039218      EBF1    0.029736      ARX   25021812   \n",
       "3  chrX    25039140  25039161      NRSE    0.019347      ARX   25021812   \n",
       "4  chrX    25039140  25039161      REST    0.029062      ARX   25021812   \n",
       "\n",
       "    geneEnd peakName  \n",
       "0  10112518        .  \n",
       "1  25034065        .  \n",
       "2  25034065        .  \n",
       "3  25034065        .  \n",
       "4  25034065        .  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"../example_data/temp.tsv\", sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## TF Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>motifStart</th>\n",
       "      <th>motifEnd</th>\n",
       "      <th>motifName</th>\n",
       "      <th>motifScore</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>peakName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>10009681</td>\n",
       "      <td>10009687</td>\n",
       "      <td>MZF1_1-4</td>\n",
       "      <td>0.395137</td>\n",
       "      <td>WWC3</td>\n",
       "      <td>9983794</td>\n",
       "      <td>10112518</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039141</td>\n",
       "      <td>25039160</td>\n",
       "      <td>REST</td>\n",
       "      <td>0.015574</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039207</td>\n",
       "      <td>25039218</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>0.029736</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039140</td>\n",
       "      <td>25039161</td>\n",
       "      <td>NRSE</td>\n",
       "      <td>0.019347</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25039140</td>\n",
       "      <td>25039161</td>\n",
       "      <td>REST</td>\n",
       "      <td>0.029062</td>\n",
       "      <td>ARX</td>\n",
       "      <td>25021812</td>\n",
       "      <td>25034065</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  motifStart  motifEnd motifName  motifScore geneName  geneStart  \\\n",
       "0  chrX    10009681  10009687  MZF1_1-4    0.395137     WWC3    9983794   \n",
       "1  chrX    25039141  25039160      REST    0.015574      ARX   25021812   \n",
       "2  chrX    25039207  25039218      EBF1    0.029736      ARX   25021812   \n",
       "3  chrX    25039140  25039161      NRSE    0.019347      ARX   25021812   \n",
       "4  chrX    25039140  25039161      REST    0.029062      ARX   25021812   \n",
       "\n",
       "    geneEnd peakName  \n",
       "0  10112518        .  \n",
       "1  25034065        .  \n",
       "2  25034065        .  \n",
       "3  25034065        .  \n",
       "4  25034065        .  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "\n",
    "expression = \"../example_data/expression/Tgfb_exp.tsv\"\n",
    "df = pd.read_csv('../example_data/temp.tsv', sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10:36:24 - GarNet: INFO - Performing linear regression for 422 transcription factor expression profiles...\n",
      "/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/statsmodels/base/model.py:1036: RuntimeWarning: invalid value encountered in true_divide\n",
      "  return self.params / self.bse\n",
      "/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater\n",
      "  return (self.a < x) & (x < self.b)\n",
      "/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less\n",
      "  return (self.a < x) & (x < self.b)\n",
      "/Users/alex/Documents/GarNet/venv/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal\n",
      "  cond2 = cond0 & (x <= self.a)\n"
     ]
    }
   ],
   "source": [
    "tfs = TF_regression(df, expression)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Occasionally, the motifScore might be 0, causing a divide by 0, causing the error above. Not to worry..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Slope</th>\n",
       "      <th>P-Value</th>\n",
       "      <th>Targets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AHR</td>\n",
       "      <td>67.850000</td>\n",
       "      <td>0.375410</td>\n",
       "      <td>[SPTBN2, EFEMP2, SYT7, STARD10, MPZL2, NRG1, S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALX4</td>\n",
       "      <td>0.744295</td>\n",
       "      <td>0.717978</td>\n",
       "      <td>[PRSS23, LSR, PFKFB3, SLC9A3R1, ACLY, ST3GAL5,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AML</td>\n",
       "      <td>-22.553067</td>\n",
       "      <td>0.094280</td>\n",
       "      <td>[FHL1, LIMS1, PDGFC, GLIS3, EPB41L4B, SPDEF, L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AP-2alphaA</td>\n",
       "      <td>6.296716</td>\n",
       "      <td>0.313251</td>\n",
       "      <td>[IRX5, PKP3, ERF, TGFB1, IL11, VIM, NRG1, TYRO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AR</td>\n",
       "      <td>-2.922142</td>\n",
       "      <td>0.414332</td>\n",
       "      <td>[PARD6A, CTNND1, TLN2, SPINT1, RHOV, PMP22, EP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ARNT</td>\n",
       "      <td>2.214483</td>\n",
       "      <td>0.366826</td>\n",
       "      <td>[AP1S2, CPT1A, EFEMP2, STARD10, ERF, CRB3, PPP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ATF1</td>\n",
       "      <td>11.183138</td>\n",
       "      <td>0.056749</td>\n",
       "      <td>[CYLD, FA2H, PARD6A, MPZL2, ERF, GEM, MYC, TYR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ATF2</td>\n",
       "      <td>-0.847601</td>\n",
       "      <td>0.434626</td>\n",
       "      <td>[CYLD, AKTIP, RELA, MPZL2, ERF, IL11, CRLF1, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ATF2:c-Jun</td>\n",
       "      <td>0.666818</td>\n",
       "      <td>0.603021</td>\n",
       "      <td>[CYLD, PARD6A, CCND1, MPZL2, ERF, CRLF1, TMEM5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ATF3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CDH1, CYLD, FA2H, PARD6A, CLCF1, ERF, ITGB1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ATF4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ERF, PRDX2, GEM, MYC, SPINT1, RHOV, NEDD4, JU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ATF6</td>\n",
       "      <td>1.495629</td>\n",
       "      <td>0.356684</td>\n",
       "      <td>[IRX3, SYT7, ERF, LYN, DUSP4, MYH9, CHSY1, PRK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ATF7IP</td>\n",
       "      <td>0.047312</td>\n",
       "      <td>0.960364</td>\n",
       "      <td>[ACSL4, CYLD, IRX3, PARD6A, CTNND1, SYT7, TAGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AhR, Arnt, HIF-1</td>\n",
       "      <td>-0.213785</td>\n",
       "      <td>0.886023</td>\n",
       "      <td>[SPTBN2, EFEMP2, SYT7, STARD10, LSR, AP1M2, CD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AhR:Arnt</td>\n",
       "      <td>0.183235</td>\n",
       "      <td>0.262017</td>\n",
       "      <td>[ACSL4, AXL, TGFB1, NFKB2, NRG1, BMF, SUZ12, A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BACH1</td>\n",
       "      <td>-1.979023</td>\n",
       "      <td>0.682020</td>\n",
       "      <td>[CDH11, PPL, PLLP, ETS1, PLEKHA7, CD44, AHNAK,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>BACH2</td>\n",
       "      <td>5.444122</td>\n",
       "      <td>0.235522</td>\n",
       "      <td>[CDH11, PPL, PLLP, SLC22A18, ETS1, AHNAK, SPTB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BARHL2</td>\n",
       "      <td>47.857580</td>\n",
       "      <td>0.952938</td>\n",
       "      <td>[RAB11FIP1, ZNF608, KITLG, LIMA1, FOXA2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BCL6</td>\n",
       "      <td>-104.446206</td>\n",
       "      <td>0.008102</td>\n",
       "      <td>[SRPX, TGFB1, VCAN, NFIB, FOXN3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>-12.109023</td>\n",
       "      <td>0.018054</td>\n",
       "      <td>[STS, SRPX, IL1RAPL1, DLG3, LOC440335, RRAD, S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>C/EBPbeta</td>\n",
       "      <td>-1.246900</td>\n",
       "      <td>0.965713</td>\n",
       "      <td>[CDH1, EPAS1, PALLD, BNC2, RUNX2, ACOT7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CACCC-binding factor</td>\n",
       "      <td>-1.993438</td>\n",
       "      <td>0.827798</td>\n",
       "      <td>[AHNAK, TCF7L2, MAFF, ELL3, TOB1, PMP22, FN1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CAMK4</td>\n",
       "      <td>0.594027</td>\n",
       "      <td>0.734611</td>\n",
       "      <td>[PRSS33, SYTL2, CTNND1, PRSS23, EHF, CCND1, RA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CD40</td>\n",
       "      <td>-8.665116</td>\n",
       "      <td>0.282167</td>\n",
       "      <td>[CYLD, RRAD, SORL1, RAB38, PRDX2, NFKB2, NRG1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CLOCK:BMAL</td>\n",
       "      <td>-2.296575</td>\n",
       "      <td>0.291624</td>\n",
       "      <td>[AP1S2, EFEMP2, MUC5B, EHF, CCND1, STARD10, CT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>COUP direct repeat 1</td>\n",
       "      <td>5.019377</td>\n",
       "      <td>0.484541</td>\n",
       "      <td>[CLCF1, BCAM, TCF7L2, NRG1, TPD52, SEPT9, DBN1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>COUP-TF, HNF-4</td>\n",
       "      <td>25.516363</td>\n",
       "      <td>0.292103</td>\n",
       "      <td>[CLCF1, TCF7L2, SEPT9, SHROOM3, NFIB, BNC2, TU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>COUPTF</td>\n",
       "      <td>-5.587996</td>\n",
       "      <td>0.190432</td>\n",
       "      <td>[AHNAK, STARD10, MKNK2, SEPT9, TRERF1, CAMK2N1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CP2/LBP-1c/LSF</td>\n",
       "      <td>7.715133</td>\n",
       "      <td>0.127056</td>\n",
       "      <td>[AHNAK, BMF, HBEGF, ZNF608, P4HA2, BTC, FOXA1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CREB, ATF</td>\n",
       "      <td>0.482868</td>\n",
       "      <td>0.511274</td>\n",
       "      <td>[CYLD, IRX3, PARD6A, SYT7, CCND1, RNF141, MPZL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>TRAF4</td>\n",
       "      <td>1.374116</td>\n",
       "      <td>0.619034</td>\n",
       "      <td>[LSR, KCTD15, DKK1, TLN2, ACLY, ABR, CCNB1, LA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>Tal-1alpha:E47</td>\n",
       "      <td>8.569764</td>\n",
       "      <td>0.468968</td>\n",
       "      <td>[ACSL4, MVP, SYT7, PRSS23, GADD45B, KCNMA1, NR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>Tal-1beta:E47</td>\n",
       "      <td>17.358153</td>\n",
       "      <td>0.259242</td>\n",
       "      <td>[ACSL4, ST3GAL2, SYT7, GADD45B, KCNMA1, STAU2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>Tal-1beta:ITF-2</td>\n",
       "      <td>20.268218</td>\n",
       "      <td>0.145046</td>\n",
       "      <td>[ACSL4, ST3GAL2, KCNMA1, NRG1, STAU2, TOX, BMP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>Tax/CREB</td>\n",
       "      <td>0.018529</td>\n",
       "      <td>0.908595</td>\n",
       "      <td>[CYLD, IRX3, PPL, ETS1, PKP3, RNF141, LSR, ERF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>UBE2N</td>\n",
       "      <td>1.265527</td>\n",
       "      <td>0.838831</td>\n",
       "      <td>[CCNF, ETS1, PLEKHA7, SPTBN2, ARNTL, MUC5B, SE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>UBP1</td>\n",
       "      <td>-19.611580</td>\n",
       "      <td>0.045062</td>\n",
       "      <td>[IDH3G, MAGED2, SYNGR3, CDH1, MVP, FA2H, CCNF,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>USF</td>\n",
       "      <td>0.251857</td>\n",
       "      <td>0.771035</td>\n",
       "      <td>[AP1S2, SBK1, CPT1A, ETS1, EFEMP2, LRP5, EXPH5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>USF1</td>\n",
       "      <td>-27.953699</td>\n",
       "      <td>0.065788</td>\n",
       "      <td>[AP1S2, TCEAL8, CPT1A, EFEMP2, MUC5B, EHF, CCN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>USF2</td>\n",
       "      <td>-0.175186</td>\n",
       "      <td>0.892555</td>\n",
       "      <td>[AP1S2, TCEAL8, SLC9A3R2, CPT1A, SLC22A18, ETS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>VBP1</td>\n",
       "      <td>-15.375917</td>\n",
       "      <td>0.028936</td>\n",
       "      <td>[SAT1, AKTIP, ERF, SSPN, FILIP1L]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>VDR:RXR</td>\n",
       "      <td>1.773976</td>\n",
       "      <td>0.754148</td>\n",
       "      <td>[SBK1, PKP3, HIF3A, ITGB3, COL1A1, SATB2, SLC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>VSX2</td>\n",
       "      <td>11.191874</td>\n",
       "      <td>0.425080</td>\n",
       "      <td>[PFKFB3, TLN2, PPM1H, BCAT1, SERPINB5, ELF3, Z...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>Vax-2</td>\n",
       "      <td>-11.300377</td>\n",
       "      <td>0.409760</td>\n",
       "      <td>[KCTD15, GGT1, TLN2, ABR, SYNPO, TLE4, ZBTB38]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>XBP1</td>\n",
       "      <td>-1.477603</td>\n",
       "      <td>0.491771</td>\n",
       "      <td>[SYT7, SCAMP2, PRKAR1A, SEPT9, CLDN7, CCNYL1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>YY1</td>\n",
       "      <td>-2.515648</td>\n",
       "      <td>0.052368</td>\n",
       "      <td>[AP1S2, DIAPH2, FHL1, SRPX, NRK, CDH1, CYLD, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>ZBTB4</td>\n",
       "      <td>2.554212</td>\n",
       "      <td>0.470766</td>\n",
       "      <td>[SYTL2, CD44, PRSS23, TCF7L2, ITGB1, CGNL1, TH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>ZBTB6</td>\n",
       "      <td>-6.036790</td>\n",
       "      <td>0.115113</td>\n",
       "      <td>[EFEMP2, AXL, PTK2B, TOM1L1, NRCAM, LRBA, EPB4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>ZEB1</td>\n",
       "      <td>-0.008059</td>\n",
       "      <td>0.998240</td>\n",
       "      <td>[FA2H, IRX3, RRAD, TGFB1I1, PLLP, ETS1, CTNND1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>ZNF143</td>\n",
       "      <td>0.915891</td>\n",
       "      <td>0.885765</td>\n",
       "      <td>[CTNND1, EFEMP2, SSH3, LMTK3, ERF, CRLF1, TMEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>ZNF217</td>\n",
       "      <td>0.071786</td>\n",
       "      <td>0.720319</td>\n",
       "      <td>[SLC9A3R2, CTSC, ERF, DMPK, ZFPM2, LOXL2, ANGP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>ZNF350</td>\n",
       "      <td>-8.416167</td>\n",
       "      <td>0.221235</td>\n",
       "      <td>[LRP5, TCF7L2, GRB7, PDLIM4, LAMB1, SCUBE3, QK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>ZNF354C</td>\n",
       "      <td>-3.305146</td>\n",
       "      <td>0.190166</td>\n",
       "      <td>[SAT1, IDH3G, ACSL4, DLG3, SYNGR3, CDH1, MVP, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>ZNF423</td>\n",
       "      <td>0.526428</td>\n",
       "      <td>0.930360</td>\n",
       "      <td>[MAGED2, CCND1, STARD10, MIF4GD, EFNB3, FOSL2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>ZNF513</td>\n",
       "      <td>-1.999460</td>\n",
       "      <td>0.773651</td>\n",
       "      <td>[SNAI2, BMP1, MRC2, SEPT9, COL1A1, PDGFC, TNFA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>alpha-CP1</td>\n",
       "      <td>5.858812</td>\n",
       "      <td>0.267957</td>\n",
       "      <td>[SAT1, DLG3, TGFB1I1, MLKL, AHNAK, H19, TM7SF2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>c-Ets-1 p54</td>\n",
       "      <td>-18.157554</td>\n",
       "      <td>0.105383</td>\n",
       "      <td>[MAGED2, AXL, CRB3, RAB11FIP1, TOX, LRRC49, SL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>c-Myc:Max</td>\n",
       "      <td>0.858400</td>\n",
       "      <td>0.540763</td>\n",
       "      <td>[AP1S2, CPT1A, EFEMP2, CCND1, STARD10, CTSF, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>myogenin / NF-1</td>\n",
       "      <td>30.746536</td>\n",
       "      <td>0.079158</td>\n",
       "      <td>[AHNAK, TCF7L2, ACLY, DLX2, ANXA6, POU3F2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>p53 decamer</td>\n",
       "      <td>413.647555</td>\n",
       "      <td>0.934799</td>\n",
       "      <td>[CD44, SORBS3, SPINT1, RHOV, EPAS1, S100A10, L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>335 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Transcription Factor       Slope   P-Value  \\\n",
       "0                     AHR   67.850000  0.375410   \n",
       "1                    ALX4    0.744295  0.717978   \n",
       "2                     AML  -22.553067  0.094280   \n",
       "3              AP-2alphaA    6.296716  0.313251   \n",
       "4                      AR   -2.922142  0.414332   \n",
       "5                    ARNT    2.214483  0.366826   \n",
       "6                    ATF1   11.183138  0.056749   \n",
       "7                    ATF2   -0.847601  0.434626   \n",
       "8              ATF2:c-Jun    0.666818  0.603021   \n",
       "9                    ATF3    0.000000       NaN   \n",
       "10                   ATF4    0.000000       NaN   \n",
       "11                   ATF6    1.495629  0.356684   \n",
       "12                 ATF7IP    0.047312  0.960364   \n",
       "13       AhR, Arnt, HIF-1   -0.213785  0.886023   \n",
       "14               AhR:Arnt    0.183235  0.262017   \n",
       "15                  BACH1   -1.979023  0.682020   \n",
       "16                  BACH2    5.444122  0.235522   \n",
       "17                 BARHL2   47.857580  0.952938   \n",
       "18                   BCL6 -104.446206  0.008102   \n",
       "19                  BRCA1  -12.109023  0.018054   \n",
       "20              C/EBPbeta   -1.246900  0.965713   \n",
       "21   CACCC-binding factor   -1.993438  0.827798   \n",
       "22                  CAMK4    0.594027  0.734611   \n",
       "23                   CD40   -8.665116  0.282167   \n",
       "24             CLOCK:BMAL   -2.296575  0.291624   \n",
       "25   COUP direct repeat 1    5.019377  0.484541   \n",
       "26         COUP-TF, HNF-4   25.516363  0.292103   \n",
       "27                 COUPTF   -5.587996  0.190432   \n",
       "28         CP2/LBP-1c/LSF    7.715133  0.127056   \n",
       "29              CREB, ATF    0.482868  0.511274   \n",
       "..                    ...         ...       ...   \n",
       "305                 TRAF4    1.374116  0.619034   \n",
       "306        Tal-1alpha:E47    8.569764  0.468968   \n",
       "307         Tal-1beta:E47   17.358153  0.259242   \n",
       "308       Tal-1beta:ITF-2   20.268218  0.145046   \n",
       "309              Tax/CREB    0.018529  0.908595   \n",
       "310                 UBE2N    1.265527  0.838831   \n",
       "311                  UBP1  -19.611580  0.045062   \n",
       "312                   USF    0.251857  0.771035   \n",
       "313                  USF1  -27.953699  0.065788   \n",
       "314                  USF2   -0.175186  0.892555   \n",
       "315                  VBP1  -15.375917  0.028936   \n",
       "316               VDR:RXR    1.773976  0.754148   \n",
       "317                  VSX2   11.191874  0.425080   \n",
       "318                 Vax-2  -11.300377  0.409760   \n",
       "319                  XBP1   -1.477603  0.491771   \n",
       "320                   YY1   -2.515648  0.052368   \n",
       "321                 ZBTB4    2.554212  0.470766   \n",
       "322                 ZBTB6   -6.036790  0.115113   \n",
       "323                  ZEB1   -0.008059  0.998240   \n",
       "324                ZNF143    0.915891  0.885765   \n",
       "325                ZNF217    0.071786  0.720319   \n",
       "326                ZNF350   -8.416167  0.221235   \n",
       "327               ZNF354C   -3.305146  0.190166   \n",
       "328                ZNF423    0.526428  0.930360   \n",
       "329                ZNF513   -1.999460  0.773651   \n",
       "330             alpha-CP1    5.858812  0.267957   \n",
       "331           c-Ets-1 p54  -18.157554  0.105383   \n",
       "332             c-Myc:Max    0.858400  0.540763   \n",
       "333       myogenin / NF-1   30.746536  0.079158   \n",
       "334           p53 decamer  413.647555  0.934799   \n",
       "\n",
       "                                               Targets  \n",
       "0    [SPTBN2, EFEMP2, SYT7, STARD10, MPZL2, NRG1, S...  \n",
       "1    [PRSS23, LSR, PFKFB3, SLC9A3R1, ACLY, ST3GAL5,...  \n",
       "2    [FHL1, LIMS1, PDGFC, GLIS3, EPB41L4B, SPDEF, L...  \n",
       "3    [IRX5, PKP3, ERF, TGFB1, IL11, VIM, NRG1, TYRO...  \n",
       "4    [PARD6A, CTNND1, TLN2, SPINT1, RHOV, PMP22, EP...  \n",
       "5    [AP1S2, CPT1A, EFEMP2, STARD10, ERF, CRB3, PPP...  \n",
       "6    [CYLD, FA2H, PARD6A, MPZL2, ERF, GEM, MYC, TYR...  \n",
       "7    [CYLD, AKTIP, RELA, MPZL2, ERF, IL11, CRLF1, T...  \n",
       "8    [CYLD, PARD6A, CCND1, MPZL2, ERF, CRLF1, TMEM5...  \n",
       "9    [CDH1, CYLD, FA2H, PARD6A, CLCF1, ERF, ITGB1, ...  \n",
       "10   [ERF, PRDX2, GEM, MYC, SPINT1, RHOV, NEDD4, JU...  \n",
       "11   [IRX3, SYT7, ERF, LYN, DUSP4, MYH9, CHSY1, PRK...  \n",
       "12   [ACSL4, CYLD, IRX3, PARD6A, CTNND1, SYT7, TAGL...  \n",
       "13   [SPTBN2, EFEMP2, SYT7, STARD10, LSR, AP1M2, CD...  \n",
       "14   [ACSL4, AXL, TGFB1, NFKB2, NRG1, BMF, SUZ12, A...  \n",
       "15   [CDH11, PPL, PLLP, ETS1, PLEKHA7, CD44, AHNAK,...  \n",
       "16   [CDH11, PPL, PLLP, SLC22A18, ETS1, AHNAK, SPTB...  \n",
       "17            [RAB11FIP1, ZNF608, KITLG, LIMA1, FOXA2]  \n",
       "18                    [SRPX, TGFB1, VCAN, NFIB, FOXN3]  \n",
       "19   [STS, SRPX, IL1RAPL1, DLG3, LOC440335, RRAD, S...  \n",
       "20            [CDH1, EPAS1, PALLD, BNC2, RUNX2, ACOT7]  \n",
       "21   [AHNAK, TCF7L2, MAFF, ELL3, TOB1, PMP22, FN1, ...  \n",
       "22   [PRSS33, SYTL2, CTNND1, PRSS23, EHF, CCND1, RA...  \n",
       "23   [CYLD, RRAD, SORL1, RAB38, PRDX2, NFKB2, NRG1,...  \n",
       "24   [AP1S2, EFEMP2, MUC5B, EHF, CCND1, STARD10, CT...  \n",
       "25   [CLCF1, BCAM, TCF7L2, NRG1, TPD52, SEPT9, DBN1...  \n",
       "26   [CLCF1, TCF7L2, SEPT9, SHROOM3, NFIB, BNC2, TU...  \n",
       "27     [AHNAK, STARD10, MKNK2, SEPT9, TRERF1, CAMK2N1]  \n",
       "28   [AHNAK, BMF, HBEGF, ZNF608, P4HA2, BTC, FOXA1,...  \n",
       "29   [CYLD, IRX3, PARD6A, SYT7, CCND1, RNF141, MPZL...  \n",
       "..                                                 ...  \n",
       "305  [LSR, KCTD15, DKK1, TLN2, ACLY, ABR, CCNB1, LA...  \n",
       "306  [ACSL4, MVP, SYT7, PRSS23, GADD45B, KCNMA1, NR...  \n",
       "307  [ACSL4, ST3GAL2, SYT7, GADD45B, KCNMA1, STAU2,...  \n",
       "308  [ACSL4, ST3GAL2, KCNMA1, NRG1, STAU2, TOX, BMP...  \n",
       "309  [CYLD, IRX3, PPL, ETS1, PKP3, RNF141, LSR, ERF...  \n",
       "310  [CCNF, ETS1, PLEKHA7, SPTBN2, ARNTL, MUC5B, SE...  \n",
       "311  [IDH3G, MAGED2, SYNGR3, CDH1, MVP, FA2H, CCNF,...  \n",
       "312  [AP1S2, SBK1, CPT1A, ETS1, EFEMP2, LRP5, EXPH5...  \n",
       "313  [AP1S2, TCEAL8, CPT1A, EFEMP2, MUC5B, EHF, CCN...  \n",
       "314  [AP1S2, TCEAL8, SLC9A3R2, CPT1A, SLC22A18, ETS...  \n",
       "315                  [SAT1, AKTIP, ERF, SSPN, FILIP1L]  \n",
       "316  [SBK1, PKP3, HIF3A, ITGB3, COL1A1, SATB2, SLC2...  \n",
       "317  [PFKFB3, TLN2, PPM1H, BCAT1, SERPINB5, ELF3, Z...  \n",
       "318     [KCTD15, GGT1, TLN2, ABR, SYNPO, TLE4, ZBTB38]  \n",
       "319  [SYT7, SCAMP2, PRKAR1A, SEPT9, CLDN7, CCNYL1, ...  \n",
       "320  [AP1S2, DIAPH2, FHL1, SRPX, NRK, CDH1, CYLD, M...  \n",
       "321  [SYTL2, CD44, PRSS23, TCF7L2, ITGB1, CGNL1, TH...  \n",
       "322  [EFEMP2, AXL, PTK2B, TOM1L1, NRCAM, LRBA, EPB4...  \n",
       "323  [FA2H, IRX3, RRAD, TGFB1I1, PLLP, ETS1, CTNND1...  \n",
       "324  [CTNND1, EFEMP2, SSH3, LMTK3, ERF, CRLF1, TMEM...  \n",
       "325  [SLC9A3R2, CTSC, ERF, DMPK, ZFPM2, LOXL2, ANGP...  \n",
       "326  [LRP5, TCF7L2, GRB7, PDLIM4, LAMB1, SCUBE3, QK...  \n",
       "327  [SAT1, IDH3G, ACSL4, DLG3, SYNGR3, CDH1, MVP, ...  \n",
       "328  [MAGED2, CCND1, STARD10, MIF4GD, EFNB3, FOSL2,...  \n",
       "329  [SNAI2, BMP1, MRC2, SEPT9, COL1A1, PDGFC, TNFA...  \n",
       "330  [SAT1, DLG3, TGFB1I1, MLKL, AHNAK, H19, TM7SF2...  \n",
       "331  [MAGED2, AXL, CRB3, RAB11FIP1, TOX, LRRC49, SL...  \n",
       "332  [AP1S2, CPT1A, EFEMP2, CCND1, STARD10, CTSF, E...  \n",
       "333         [AHNAK, TCF7L2, ACLY, DLX2, ANXA6, POU3F2]  \n",
       "334  [CD44, SORBS3, SPINT1, RHOV, EPAS1, S100A10, L...  \n",
       "\n",
       "[335 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### If you provide an output_dir to `TF_regression`, you can see each of the scatterplots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
